src/combinatorics/additive/behrend.lean

```diff
@@ -0,0 +1,192 @@
+
+/-!
+### Turning the sphere into a Salem-Spencer set
+
+We define `behrend.sphere`, the intersection of the $$L^2$$ sphere with the positive quadrant of
+integer points. Because the $$L^2$$ closed ball is strictly convex, the $$L^2$$ sphere and
+`behrend.sphere` are Salem-Spencer (`add_salem_spencer_sphere`). Then we can turn this set in
+`fin n → ℕ` into a set in `ℕ` using `behrend.map`, which preserves `add_salem_spencer` because it is
+an additive monoid homomorphism.
+-/
+
+/-- The box `{0, ..., d - 1}^n` as a finset. -/
+def box (n d : ℕ) : finset (fin n → ℕ) := fintype.pi_finset $ λ _, range d
+
+lemma mem_box : x ∈ box n d ↔ ∀ i, x i < d := by simp only [box, fintype.mem_pi_finset, mem_range]
+
+@[simp] lemma card_box : (box n d).card = d ^ n := by simp [box]
+
+/-- The intersection of the sphere of radius `sqrt k` with the integer points in the positive
+quadrant. -/
+def sphere (n d k : ℕ) : finset (fin n → ℕ) := (box n d).filter $ λ x, ∑ i, x i^2 = k
+
+lemma sphere_zero : sphere n d 0 ⊆ 0 :=
+λ x, by simp [sphere, function.funext_iff] {contextual := tt}
+
+lemma sphere_subset_box : sphere n d k ⊆ box n d := filter_subset _ _
+
+lemma norm_of_mem_sphere {n d k : ℕ} {x : fin n → ℕ} (hx : x ∈ sphere n d k) :
+  ∥@id (euclidean_space ℝ (fin n)) (coe ∘ x : fin n → ℝ)∥ = sqrt k :=
+begin
+  rw euclidean_space.norm_eq,
+  congr',
+  change ∑ (i : fin n), ∥(x i : ℝ)∥ ^ 2 = k,
+  simp_rw [norm_coe_nat, ←cast_pow, ←cast_sum, (mem_filter.1 hx).2],
+end
+
+/-- The map that appears in Behrend's bound on Roth numbers. -/
+@[simps] def map (d : ℕ) : (fin n → ℕ) →+ ℕ :=
+{ to_fun := λ a, ∑ i, a i * d ^ (i : ℕ),
+  map_zero' := by simp_rw [pi.zero_apply, zero_mul, sum_const_zero],
+  map_add' := λ a b, by simp_rw [pi.add_apply, add_mul, sum_add_distrib] }
+
+@[simp] lemma map_zero (d : ℕ) (a : fin 0 → ℕ) : map d a = 0 := by simp [map]
+
+lemma map_succ (a : fin (n + 1) → ℕ) : map d a = a 0 + (∑ x : fin n, a x.succ * d ^ (x : ℕ)) * d :=
+by simp [map, fin.sum_univ_succ, pow_succ', ←mul_assoc, ←sum_mul]
+
+lemma map_succ' (a : fin (n + 1) → ℕ) : map d a = a 0 + map d (a ∘ fin.succ) * d := map_succ _
+
+lemma map_monotone (d : ℕ) : monotone (map d : (fin n → ℕ) → ℕ) :=
+λ x y h, by { dsimp, exact sum_le_sum (λ i _, nat.mul_le_mul_right _ $ h i) }
+
+lemma map_mod (a : fin n.succ → ℕ) : map d a % d = a 0 % d :=
+by rw [map_succ, nat.add_mul_mod_self_right]
+
+lemma map_eq_iff {x₁ x₂ : fin n.succ → ℕ} (hx₁ : ∀ i, x₁ i < d) (hx₂ : ∀ i, x₂ i < d) :
+  map d x₁ = map d x₂ ↔ x₁ 0 = x₂ 0 ∧ map d (x₁ ∘ fin.succ) = map d (x₂ ∘ fin.succ) :=
+begin
+  refine ⟨λ h, _, λ h, by rw [map_succ', map_succ', h.1, h.2]⟩,
+  have : x₁ 0 = x₂ 0,
+  { rw [←mod_eq_of_lt (hx₁ _), ←map_mod, ←mod_eq_of_lt (hx₂ _), ←map_mod, h] },
+  rw [map_succ, map_succ, this, add_right_inj, mul_eq_mul_right_iff] at h,
+  exact ⟨this, h.resolve_right (pos_of_gt (hx₁ 0)).ne'⟩,
+end
+
+lemma map_inj_on : {x : fin n → ℕ | ∀ i, x i < d}.inj_on (map d) :=
```

Comment:
Any reason not to state it this way?
```suggestion
lemma map_inj_on : (box n d : set _).inj_on (map d) :=
```

Reply:
Definitional equality.

Reply:
This has worse definitional equalities and I do not actually have `x ∈ box n d` hypotheses at the place of use.

Reply:
No strong feelings here but I do note that all it would take use Eric's proposed statement at the place of use would be to change:
```lean
rw set.add_subset_iff,
```
to
```lean
simp_rw [set.add_subset_iff, mem_coe, mem_box],
```

Reply:
GIven that this is local to the proof, I won't bother.